graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/AbstractGraphqlService.java

```diff
@@ -0,0 +1,219 @@
+                }
+            }));
+        }
+
+        return unsupportedMediaType();
+    }
+
+    /**
+     * Handles a {@link GraphqlRequest}.
+     */
+    protected abstract HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest req)
+            throws Exception;
+
+    private static Map<String, Object> toMap(@Nullable String value) throws JsonProcessingException {
+        if (Strings.isNullOrEmpty(value)) {
+            return ImmutableMap.of();
+        }
+        return parseJsonString(value);
+    }
+
+    private static Map<String, Object> toMap(@Nullable Object maybeMap) {
+        if (maybeMap == null) {
+            return ImmutableMap.of();
+        }
+
+        if (maybeMap instanceof Map) {
+            final Map<?, ?> variablesMap = (Map<?, ?>) maybeMap;
+            if (variablesMap.isEmpty()) {
+                return ImmutableMap.of();
+            }
+
+            final ImmutableMap.Builder<String, Object> builder =
+                    ImmutableMap.builderWithExpectedSize(variablesMap.size());
+            variablesMap.forEach((k, v) -> builder.put(String.valueOf(k), v));
+            return builder.build();
+        } else {
+            throw new IllegalArgumentException("Unknown parameter type variables");
+        }
+    }
+
+    private static Map<String, Object> parseJsonString(String content) throws JsonProcessingException {
+        return mapper.readValue(content, JSON_MAP);
+    }
+
+    private static HttpResponse unsupportedMediaType() {
+        return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE,
+                               MediaType.PLAIN_TEXT,
+                               "Unsupported media type. Only JSON compatible types and " +
+                               "application/graphql are supported.");
+    }
+
+    /**
+     * Returns the negotiated {@link MediaType}. {@link MediaType#JSON} and {@link MediaType#GRAPHQL_JSON}
+     * are commonly used for the Content-Type of a GraphQL response.
+     * If {@link HttpHeaderNames#ACCEPT} is not specified, {@link MediaType#GRAPHQL_JSON} is used by default.
+     *
+     * <p>Note that the negotiated {@link MediaType} could not be used by the implementation of
+     * {@link AbstractGraphqlService} which may choose to respond in one of several ways
+     * specified the
+     * <a href="https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#body">
+     * specification</a>.
+     */
+    @Nullable
+    private static MediaType produceType(RequestHeaders headers) {
+        final List<MediaType>  acceptTypes = headers.accept();
+        if (acceptTypes.isEmpty()) {
+            // If there is no Accept header in the request, the response MUST include
+            // a Content-Type: application/graphql+json header
+            return MediaType.GRAPHQL_JSON;
+        }
+
+        for (MediaType accept : acceptTypes) {
+            if (accept.is(MediaType.GRAPHQL_JSON) || accept.is(MediaType.JSON)) {
```

Comment:
How about handling `"application/graphql"` as well? 
https://graphql.org/learn/serving-over-http/#post-request

Reply:
`application/graphql` format is different from JSON. This `MediaType` is only used for a response that must be serialized as a JSON.